src/scanners.rs

```diff
@@ -456,6 +456,143 @@ pub fn calc_indent(text: &str, max: usize) -> (usize, usize) {
     (i, spaces)
 }
 
+pub struct RefDefScan<'a> {
+    pub(crate) bytecount: usize,
+    pub(crate) label: &'a str,
+    pub(crate) dest: &'a str,
+    pub(crate) title: Option<&'a str>
+}
+
+pub fn scan_refdef(data: &str) -> Option<RefDefScan> {
+    let mut i = scan_whitespace_no_nl(data);
+
+    // defs can only be indented up to three spaces
+    if i > 3 {
+        return None;
+    }
+
+    // scan label
+    if scan_ch(&data[i..], b'[') == 0 {
+        return None;
+    }
+    i += 1;
+    let label_length = scan_while(&data[i..], |c| c != b']');
+    let label = &data[i..(i + label_length)];
+    i += label_length;
+    // FIXME: out of bounds possibly?
+    if scan_ch(&data[i..], b']') == 0 || scan_ch(&data[(i + 1)..], b':') == 0 {
+        return None;
+    }
+    i += 2;
+
+    // whitespace between label and url (including up to one newline)
+    let mut newlines = 0;
+    i += scan_while(&data[i..], |c| {
+        if c == b'\n' {
+            newlines += 1;
```

Comment:
*Probably* want to break early on `newlines > 1`, but that's a minor optimization.